Prepare for release 0.9.0-preview.0

toasteater opened the pull request:
~~Issue number squatting~~ Preparation PR for the 0.9.0 preview release!

Version 0.9 contains a lot of large interface changes that are intended to be solutions to long-standing problems in the API. A lot of good contributions have been merged in the past months, and they are ready for some real-world testing! Since the changes are rather large, we'll release a preview version to weed out any problems before actually committing to the version bump. If no problems are found, this will then proceed to become 0.9!

Draft status: we'll want to clear the other issues in the 0.9 milestone before merging this.

toasteater commented:
Updated README and CHANGELOG. The 0.9 milestone is almost done so I think this is ready for review!

karroffel commented on CHANGELOG.md:
```diff
@@ -7,6 +7,76 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+- The typed arrays are unified into a generic `TypedArray` type. The old names remain as type aliases.
+
+- Moved generated bindings into the `gdnative::api` module, making the top-level namespace easier to navigate.
+
```
I'd add that it's now possible to create custom bindings crates without having to fork the repository since `Api::new()` takes all the data as an argument now? #480

toasteater replied:
Added!

toasteater commented:
Removed #514 from known issues. Removed the information about 4.0 support from known issues as well, since it's mentioned in README.

toasteater commented:
Rebased and updated CHANGELOG and README. I have trimmed most example section in favor of the tutorial in the book. Please tell me if this is a good direction!

karroffel commented on README.md:
```diff
@@ -20,10 +18,14 @@ We are serious about engine compatibility. We are committed to keeping compatibi
 
 The current minimum compatible version, with `api.json` replacement, is Godot 3.1.1-stable. Changes to this will be considered a breaking change, and will be called out in the release notes.
 
+The bindings does *not* support Godot 4.0 (`master` branch) currently.
```
```suggestion
The bindings do *not* support Godot 4.0 (`master` branch) currently.
```

karroffel commented on README.md:
```diff
@@ -43,10 +45,10 @@ crate-type = ["cdylib"]
 The bindings are currently generated from the API description of Godot 3.2 by default. To use the bindings with another version or a custom build, one currently has to use the bindings as a local dependency:
```
I guess this can be changed too as full local clone is technically not needed anymore? With a section in the book about custom bindings this should be obsolete?

toasteater replied:
Opened a PR to add a guide to the book: https://github.com/godot-rust/book/pull/7

toasteater commented:
Planning to finish the migration guide and publish the version tomorrow!

karroffel commented on README.md:
```diff
@@ -67,99 +52,34 @@ classes, as well as providing custom functionality by exposing Rust types as *Na
-    // methods MUST have `owner: BaseClass` as their second arguments,
-    // before all other arguments in the signature.
-    #[export]
+#[export]
```
This should be indented as before

toasteater replied:
Fixed!

toasteater commented:
bors r+

toasteater commented:
To authors of third-party resources:

We're about to release version 0.9, which contains many breaking changes that make it incompatible with older resources for 0.8.x. If you're willing to update your tutorials / projects for 0.9 ([preview release](https://crates.io/crates/gdnative/0.9.0-preview.0)), please let us know. Third-party resources that are not updated will be removed from README at 0.9 release.

Cc: @tom-leys @hagsteel @paytonrules @tommywalkie

hagsteel commented:
I've updated my guide.

paytonrules commented:
There’s a lot here, and my part two is still in progress.

What I may do is go ahead and write a part three that’s - “Updating to version 0.9”.
On Jul 10, 2020, 11:44 AM -0500, toasteater <notifications@github.com>, wrote:
> To authors of third-party resources:
> We're about to release version 0.9, which contains many breaking changes that make it incompatible with older resources for 0.8.x. If you're willing to update your tutorials / projects for 0.9 (preview release), please let us know. Third-party resources that are not updated will be removed from README at 0.9 release.
> Cc: @tom-leys @hagsteel @paytonrules @tommywalkie
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

tommywalkie commented:
@toasteater @karroffel Thanks for the update ! Gotta look at CHANGELOG and update my guide 👍

paytonrules commented:
As the book is in progress it could really use more explanation and examples around Ref/TRef and assume_safe and all it's variations. I'll give an example for clarification:

I'm moving my 0.8 code over to 0.9, and as part of my code I had a struct that held an Optional sprite, that's been changed to fit:

```
pub struct Enemy {
    sprite: Option<Ref<Sprite>>,
...
}
````

When I create the sprite I give it a name and assign it a texture at random from the resources. Unfortunately the entire function is still unsafe. At this time I create a mutable shared ref, and assume it's safe. After setting the texture I add the child, and then after it's added I save a reference - making sure to `claim` it. This compiles, huzzah, but I'm 99% sure it's wrong.

```
    #[export]
    unsafe fn _enter_tree(&mut self, owner: &Node2D) {
        /* ...  */ 
        let mut sprite = Sprite::new().into_shared().assume_safe();
        let sprite_name = /* irrelevant */
        let mut resource_loader = ResourceLoader::godot_singleton();
        let texture = /* irrelevant *

        sprite.set_texture(texture);
        owner.add_child(sprite, false);
        self.sprite = Some(sprite.claim());
    }
```
